### PR TITLE
Issue #839 - Resid selection with icodes

### LINKED ISSUE #839: Inserted residues in PDB files cannot be individually selected
### Expected behaviour

The PDB format includes an iCode (insertion code) field which extends residue IDs to give inserted residues unique identifiers. Three residues inserted between resid 163 and 164 would become 163A, 163B and 163C. This allows insertion of residues while maintaining consistency with the original numbering.

Using PyMol the following selections work as intended:

```
PyMOL>fetch 1osm
PyMOL>select sele, resid 163; select sele resid 163A
 Selector: selection "sele" defined with 18 atoms.
 Selector: selection "sele" defined with 39 atoms.
```

I would expect those selection strings to produce the same results in MDAnalysis.
### Actual behaviour

`resid 163` actually selects residue 163 and 10 inserted residues (A through J) in the PDB deposition.

`resid 163A` results in an error : 'Failed to parse number: 163A'.
### Code to reproduce the behaviour

Bash (for anyone in a hurry):

``` bash
$ python -i -c 'import MDAnalysis as mda; import urllib; downloader = urllib.URLopener(); downloader.retrieve("http://files.rcsb.org/download/1osm.pdb", "1osm.pdb"); u = mda.Universe("1osm.pdb")'
>>> u.select_atoms("resid 163")
<AtomGroup with 204 atoms>
>>> u.select_atoms("resid 163A")
MDAnalysis.exceptions.SelectionError: Selection failed: 'Failed to parse number: 163A'
```

Or Python:

``` python
import MDAnalysis as mda
import urllib; downloader = urllib.URLopener()
downloader.retrieve("http://files.rcsb.org/download/1osm.pdb", "1osm.pdb")
u = mda.Universe("1osm.pdb")
print u.select_atoms("resid 163")
print u.select_atoms("resid 163A")
```
### Currently version of MDAnalysis:

0.14.0

### PR DESCRIPTION
**Pull request by richardjgowers**
Fixes #839

Changes made in this Pull Request:
 - Resid selections can now be suffixed by an icode

Eg:

```python
u.select_atoms('resid 163A')
# is the same as
u.select_atoms('resid 163 and icode A')
```

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Comment by richardjgowers** [comment]
I got a little stuck on this, and had a question.  If there are `icodes` in the topology, and for example resid `163`, `163A` and `163B`, what should `select_atoms('resid 163')` select?  All three residues with resid 163, or just the residue without an icode?  I'm leaning towards the latter.

**Comment by kain88-de** [comment]
I would lean to the later as well. But people might expect the one with the highest occupancy to show up.

See:
http://www.wwpdb.org/documentation/file-format-content/format33/sect9.html#ATOM

**Comment by orbeckst** [comment]
@kain88-de I don't think that iCode has anything to do with occupancy (altLoc does), iCode allows you to insert a residue in a mutant while still using the wild type residue numbering. I think the solution proposed by @richardjgowers to use an exact match is the correct one.

**Comment by kain88-de** [comment]
@orbeckst yes you are right I confused it with the altLoc specifier.

**Comment by richardjgowers** [comment]
Ok so this is good to go now.

File: package/MDAnalysis/core/groups.py [hunk 1]
```diff
@@ -995,6 +995,11 @@ def select_atoms(self, sel, *othersel, **selgroups):
                 range consists of two numbers separated by a colon (inclusive)
                 such as ``resid 1:5``. A residue number ("resid") is taken
                 directly from the topology.
+                If icodes are present in the topology, then these will be
+                taken into account.  Ie 'resid 163B' will only select resid
+                163 with icode B while 'resid 163' will select only residue 163.
```
Comment (kain88-de) [inline package/MDAnalysis/core/groups.py]: I thought we always select the first residue with number 163 now. With what ever icode that happens to be. So the docs could say that 'resid 163' will select the first residue in the topology. While one can also specify icodes to select specific residues with the same number.

But what about PDB files like this

```
ATOM     34  CA AARG A   2C      12.353  85.696  94.456  0.50 36.67           C
ATOM     35  CA BARG A   2A      12.333  85.862  95.041  0.50 36.42           C
```

Where the first icode lines aren't sorted alphabetically? Which atom would `resid 2` select in this case?

**Comment by kain88-de** [comment]
Also what happens in the following case
```
ATOM     34  CA AARG A   2       12.353  85.696  94.456  0.50 36.67           C
ATOM     35  CA BARG A   2A      12.333  85.862  95.041  0.50 36.42           C
```

and 

```
ATOM     34  CA AARG A   2C      12.353  85.696  94.456  0.50 36.67           C
ATOM     35  CA BARG A   2       12.333  85.862  95.041  0.50 36.42           C
```

What would `resid 2` select?

**Comment by richardjgowers** [comment]
In the `2 2A` and `2C 2` case, `resid 2` always selects the residue without an insertion code.  

In general the order / sorting of values won't matter when doing selections.  Eg if you're selecting `resid 2A-7` from `1 2 2A 2B 12 13 14 2C 6 7` you will still select `2A 2B 2C 6 7` despite the values in the middle.

**Pull request merged**